File: bench/WaveletKernel.swift
// Swift port of the representative wavelet kernel (mirror of bench/wavelet_kernel.cpp).
// Two implementations:
//   naive  - idiomatic [Double] subscripting (bounds-checked under -O)
//   tuned  - withUnsafeMutableBufferPointer, raw pointer arithmetic
// Compile once with -O and once with -Ounchecked to see the bounds-check tax.
//
// Build: swiftc -O            bench/WaveletKernel.swift -o bench/wk_swift_O
//        swiftc -Ounchecked   bench/WaveletKernel.swift -o bench/wk_swift_Ou

import Foundation

let H: [Double] = [
    0.054415842, -0.312871590, 0.675630736, -0.585354684,
    -0.015829105, 0.284015543, -0.000472485, -0.128747427,
    0.017369301, 0.044088254, -0.013981028, -0.008746094,
    0.004870353, -0.000391740, -0.000675449, 0.000117477,
]
let G: [Double] = (0..<16).map { k in (k & 1 == 1 ? -1.0 : 1.0) * H[15 - k] }

// ---- naive: idiomatic array subscripting ----
func fwt2dNaive(_ a: inout [Double], _ tmp: inout [Double], _ W: Int, _ levels: Int) {
    for lvl in 0..<levels {
        let n = W >> lvl
        let half = n >> 1
        for r in 0..<n { // rows
            let rb = r * W
            for i in 0..<n { tmp[i] = a[rb + i] }
            for i in 0..<half {
                var av = 0.0, dv = 0.0
                let base = 2 * i
                for k in 0..<16 {
                    var idx = base + k; if idx >= n { idx -= n }
                    let v = tmp[idx]
                    av += H[k] * v; dv += G[k] * v
                }
                a[rb + i] = av
                a[rb + half + i] = dv
            }
        }
        for c in 0..<n { // columns
            for i in 0..<n { tmp[i] = a[i * W + c] }
            for i in 0..<half {
                var av = 0.0, dv = 0.0
                let base = 2 * i
                for k in 0..<16 {
                    var idx = base + k; if idx >= n { idx -= n }
                    let v = tmp[idx]
                    av += H[k] * v; dv += G[k] * v
                }
                a[i * W + c] = av
                a[(half + i) * W + c] = dv
            }
        }
    }
}

// ---- tuned: unsafe buffer pointers ----
func fwt2dTuned(_ a: inout [Double], _ tmp: inout [Double], _ W: Int, _ levels: Int) {
    H.withUnsafeBufferPointer { hb in
    G.withUnsafeBufferPointer { gb in
    a.withUnsafeMutableBufferPointer { ab in
    tmp.withUnsafeMutableBufferPointer { tb in
        let h = hb.baseAddress!, g = gb.baseAddress!
        let A = ab.baseAddress!, T = tb.baseAddress!
        for lvl in 0..<levels {
            let n = W >> lvl
            let half = n >> 1
            for r in 0..<n {
                let rb = r * W
                for i in 0..<n { T[i] = A[rb + i] }
                for i in 0..<half {
                    var av = 0.0, dv = 0.0
                    let base = 2 * i
                    for k in 0..<16 {
                        var idx = base + k; if idx >= n { idx -= n }
                        let v = T[idx]
                        av += h[k] * v; dv += g[k] * v
                    }
                    A[rb + i] = av
                    A[rb + half + i] = dv
                }
            }
            for c in 0..<n {
                for i in 0..<n { T[i] = A[i * W + c] }
                for i in 0..<half {
                    var av = 0.0, dv = 0.0
                    let base = 2 * i
                    for k in 0..<16 {
                        var idx = base + k; if idx >= n { idx -= n }
                        let v = T[idx]
                        av += h[k] * v; dv += g[k] * v
                    }
                    A[i * W + c] = av
                    A[(half + i) * W + c] = dv
                }
            }
        }
    }}}}
}

func bench(_ label: String, _ fn: (inout [Double], inout [Double], Int, Int) -> Void) {
    let W = 512, levels = 5, iters = 100
    let count = W * W
    var input = [Double](repeating: 0, count: count)
    for y in 0..<W {
        for x in 0..<W {
            input[y * W + x] = Double((x * 131 + y * 977) % 251) / 251.0
        }
    }
    var work = [Double](repeating: 0, count: count)
    var tmp = [Double](repeating: 0, count: W)

    for i in 0..<count { work[i] = input[i] }
    fn(&work, &tmp, W, levels) // warm-up

    var times = [Double]()
    var checksum = 0.0
    for it in 0..<iters {
        for i in 0..<count { work[i] = input[i] } // untimed refill
        let t0 = DispatchTime.now().uptimeNanoseconds
        fn(&work, &tmp, W, levels)
        let t1 = DispatchTime.now().uptimeNanoseconds
        times.append(Double(t1 - t0) / 1e6)
        if it == iters - 1 { for v in work { checksum += v } }
    }
    times.sort()
    print(String(format: "%@ %7.3f ms/transform   checksum=%.6f",
                 label, times[times.count / 2], checksum))
}

#if OUNCHECKED
let mode = "-Ounchecked"
#else
let mode = "-O        "
#endif
bench("Swift naive (\(mode))", fwt2dNaive)
bench("Swift tuned (\(mode))", fwt2dTuned)

File: bench/wavelet_kernel.cpp
// Language head-to-head: a representative wavelet kernel (same shape as GLIC's
// Symlet8 forward transform — 16-tap separable periodic convolution + downsample,
// multi-level Mallat). NOT bit-identical to GLIC; matched in arithmetic shape so
// the only variable is C++ vs Swift codegen. Checksum lets us confirm the Swift
// port computes the identical result.
//
// Build: clang++ -std=c++17 -O2 -DNDEBUG bench/wavelet_kernel.cpp -o bench/wk_cpp

#include <chrono>
#include <cstdio>
#include <vector>

using clk = std::chrono::steady_clock;

static const double H[16] = {
    0.054415842, -0.312871590, 0.675630736, -0.585354684,
    -0.015829105, 0.284015543, -0.000472485, -0.128747427,
    0.017369301, 0.044088254, -0.013981028, -0.008746094,
    0.004870353, -0.000391740, -0.000675449, 0.000117477};
static double G[16];

// 1D forward step: src contiguous (stride 1), dst with stride `ds`.
// First half = approximation, second half = detail. Periodic (n >= 16).
static inline void fwt1d(const double* src, double* dst, int n, int ds) {
    const int half = n >> 1;
    for (int i = 0; i < half; ++i) {
        double a = 0.0, d = 0.0;
        const int base = 2 * i;
        for (int k = 0; k < 16; ++k) {
            int idx = base + k;
            if (idx >= n) idx -= n;
            const double v = src[idx];
            a += H[k] * v;
            d += G[k] * v;
        }
        dst[i * ds] = a;
        dst[(half + i) * ds] = d;
    }
}

static void fwt2d(std::vector<double>& a, std::vector<double>& tmp, int W, int levels) {
    for (int lvl = 0; lvl < levels; ++lvl) {
        const int n = W >> lvl;
        for (int r = 0; r < n; ++r) { // rows
            double* row = a.data() + (size_t)r * W;
            for (int i = 0; i < n; ++i) tmp[i] = row[i];
            fwt1d(tmp.data(), row, n, 1);
        }
        for (int c = 0; c < n; ++c) { // columns
            for (int i = 0; i < n; ++i) tmp[i] = a[(size_t)i * W + c];
            fwt1d(tmp.data(), a.data() + c, n, W);
        }
    }
}

int main() {
    for (int k = 0; k < 16; ++k) G[k] = ((k & 1) ? -1.0 : 1.0) * H[15 - k];

    const int W = 512, levels = 5, iters = 100;
    const size_t count = (size_t)W * W;

    std::vector<double> input(count), work(count), tmp(W);
    for (int y = 0; y < W; ++y)
        for (int x = 0; x < W; ++x)
            input[(size_t)y * W + x] = (double)((x * 131 + y * 977) % 251) / 251.0;

    work = input;
    fwt2d(work, tmp, W, levels); // warm-up

    std::vector<double> times;
    double checksum = 0.0;
    for (int it = 0; it < iters; ++it) {
        for (size_t i = 0; i < count; ++i) work[i] = input[i]; // untimed refill
        auto t0 = clk::now();
        fwt2d(work, tmp, W, levels);
        auto t1 = clk::now();
        times.push_back(std::chrono::duration<double, std::milli>(t1 - t0).count());
        if (it == iters - 1) {
            for (size_t i = 0; i < count; ++i) checksum += work[i];
        }
    }
    std::sort(times.begin(), times.end());
    printf("C++   (-O2)        %7.3f ms/transform   checksum=%.6f\n",
           times[times.size() / 2], checksum);
    return 0;
}
